### PR TITLE
Add pixelbuf for accelerating NeoPixel operations

### PR DESCRIPTION
**Pull request by rhooper**
This PR adds a PixelBuf class that accelerates byteorder and brightness operations and will have an associated PR for https://github.com/adafruit/Adafruit_CircuitPython_NeoPixel
Relates to issue https://github.com/adafruit/circuitpython/issues/884

Further work is needed to:
- disable for builds it doesn't fit into or make sense for
- add support for dotstars
- optimize it further (feedback please!)

### CODE REVIEW

File: shared-bindings/pixelbuf/PixelBuf.c [hunk 1]
```diff
@@ -0,0 +1,506 @@
...
+MP_DEFINE_CONST_FUN_OBJ_1(pixelbuf_pixelbuf_get_byteorder_obj, pixelbuf_pixelbuf_obj_get_byteorder);
+
+const mp_obj_property_t pixelbuf_pixelbuf_byteorder_obj = {
+    .base.type = &mp_type_property,
+    .proxy = {(mp_obj_t)&pixelbuf_pixelbuf_get_byteorder_obj,
+              (mp_obj_t)&mp_const_none_obj,
+              (mp_obj_t)&mp_const_none_obj},
+};
+
+//|   .. attribute:: len
```
Comment (tannewt) [inline shared-bindings/pixelbuf/PixelBuf.c]: Just use len() through the unary op. Its confusing to have multiple ways of doing things.

File: shared-bindings/pixelbuf/PixelBuf.c [hunk 2]
```diff
@@ -0,0 +1,506 @@
...
+MP_DEFINE_CONST_FUN_OBJ_1(pixelbuf_pixelbuf_get_rawbuf_obj, pixelbuf_pixelbuf_obj_get_rawbuf);
+
+const mp_obj_property_t pixelbuf_pixelbuf_rawbuf_obj = {
+    .base.type = &mp_type_property,
+    .proxy = {(mp_obj_t)&pixelbuf_pixelbuf_get_rawbuf_obj,
+              (mp_obj_t)&mp_const_none_obj,
+              (mp_obj_t)&mp_const_none_obj},
+};
+
+//|   .. attribute:: size
```
Comment (tannewt) [inline shared-bindings/pixelbuf/PixelBuf.c]: Why do you need this? Can't it be computed from the len and pixel order?

File: shared-bindings/pixelbuf/PixelBuf.c [hunk 3]
```diff
@@ -0,0 +1,506 @@
...
+//|
+//| .. class:: PixelBuf(size, buf, byteorder=BGR, bpp=3)
+//|
+//|   Create a PixelBuf object of the specified size, byteorder, and bits per pixel.
+//|
+//|   :param ~int size: Number of pixels
+//|   :param ~bytearray buf: Bytearray to store pixel data in
+//|   :param ~pixelbuf.BGR: Byte order
+//|   :param ~int: Bytes per pixel
+//|
```
Comment (tannewt) [inline shared-bindings/pixelbuf/PixelBuf.c]: This should go at the top of this file.

File: shared-bindings/pixelbuf/PixelBuf.c [hunk 4]
```diff
@@ -0,0 +1,506 @@
...
+                        pixelstart[j] = (pixelstart[j] * self->brightness);
+                    }
+                }
+                return mp_const_none;
+            }
+        }
+    }
+}
+
+void pixelbuf_set_pixel_int(uint8_t *buf, mp_int_t value, uint byteorder, uint bpp) {
```
Comment (tannewt) [inline shared-bindings/pixelbuf/PixelBuf.c]: These five functions should be in shared-module instead. This file should only contain the conversion from Python to C.

File: shared-bindings/pixelbuf/PixelBuf.c [hunk 5]
```diff
@@ -0,0 +1,506 @@
...
+                } else {
+                    destbuf = self->buf;
+                }
+                if (MP_OBJ_IS_TYPE(value, &mp_type_array) ||
+                        MP_OBJ_IS_TYPE(value, &mp_type_bytearray) ||
+                        MP_OBJ_IS_TYPE(value, &mp_type_memoryview) ||
+                        MP_OBJ_IS_TYPE(value, &mp_type_bytes)) {
+                    mp_get_buffer_raise(value, &bufinfo, MP_BUFFER_READ);
+                    if (1 != mp_binary_get_size('@', bufinfo.typecode, NULL)) {
+                        mp_raise_ValueError("Array values should be single bytes.");
```
Comment (tannewt) [inline shared-bindings/pixelbuf/PixelBuf.c]: Why? Wouldn't it make more sense as an array of 32 bit values, one for each color? That or a list of tuples.

File: shared-bindings/pixelbuf/PixelBuf.c [hunk 6]
```diff
@@ -0,0 +1,506 @@
...
+
+void pixelbuf_recalculate_brightness(pixelbuf_pixelbuf_obj_t *self) {
+    for (uint i = 0; i < self->bytes; i++) {
+       self->buf[i] = self->rawbuf[i] * self->brightness;
+    }
+}
+
+//|   .. attribute:: buf
+//|
+//|     The bytearray of pixel data after brightness adjustment
```
Comment (tannewt) [inline shared-bindings/pixelbuf/PixelBuf.c]: Why is this needed externally?

File: shared-bindings/pixelbuf/PixelBuf.c (hunk 1 above)
Comment (rhooper) [reply shared-bindings/pixelbuf/PixelBuf.c]: Agreed

File: shared-bindings/pixelbuf/PixelBuf.c (hunk 2 above)
Comment (rhooper) [reply shared-bindings/pixelbuf/PixelBuf.c]: I honestly have no good explanation for this one.  I was probably thinking along the lines of "if i'm working with bytes..." But I'm about to purge that code too.

File: shared-bindings/pixelbuf/PixelBuf.c (hunk 6 above)
Comment (rhooper) [reply shared-bindings/pixelbuf/PixelBuf.c]: rawbuf is the one that's not needed externally.  buf is the bytearray needed for sending to neopixels.

**Comment by tannewt** [comment]
@rhooper Are you waiting on me for this? It'd be awesome to get it in 4.0.0. Thanks!

**Comment by rhooper** [comment]
No, i'm not waiting on you, i'm waiting on my time to free up.
I did start on the classes for RGBLED and a Pure python pixelbuf
https://github.com/rhooper/Adafruit_CircuitPython_RGBLED
from which we can make NeoPixel and Dotstar (and others) inherit.

**Comment by rhooper** [comment]
This is ready for another look, along with the code in:
- https://github.com/rhooper/Adafruit_CircuitPython_RGBLED/blob/master/adafruit_dotstar.py
- https://github.com/rhooper/Adafruit_CircuitPython_RGBLED/blob/master/neopixel.py
- https://github.com/rhooper/Adafruit_CircuitPython_RGBLED/blob/master/pypixelbuf.py

File: shared-bindings/_pixelbuf/PixelBuf.c [hunk 7]
```diff
@@ -0,0 +1,506 @@
...
+            pixelbuf_set_pixel(self->buf + offset, self->two_buffers ? self->rawbuf + offset : NULL, 
+                self->brightness, value, &self->byteorder, self->dotstar_mode);
+            if (self->auto_write)
+                call_write_function(self);
+            return mp_const_none;
+        }
+    }
+}
+
+//|   .. method:: fill_wheel(start=0, step=1)
```
[the remark this answers is not in the document]
Comment (rhooper) [reply shared-bindings/_pixelbuf/PixelBuf.c]: Yeah, it is.  Should we add a standard way to let people easily add custom C functions and put this in there as an example.  Maybe some other nifty examples that don't belong in the core.  

The learning curve to add modules and functions is pretty high right now, but having a `custom` module that's present in self-built CPy could ease that.  That gives us an easier path for people to contribute things like FFTs and the like.

Then someday we can get dynamic loading :-)

File: ports/atmel-samd/boards/circuitplayground_express_crickit/mpconfigboard.mk [hunk 8]
```diff
@@ -9,6 +9,8 @@ EXTERNAL_FLASH_DEVICE_COUNT = 2
 EXTERNAL_FLASH_DEVICES = "S25FL216K, GD25Q16C"
 # Turn off longints for Crickit build to make room for additional frozen libs.
 LONGINT_IMPL = NONE
+# Disable pixelbuf to save room
```
[the remark this answers is not in the document]
Comment (rhooper) [reply ports/atmel-samd/boards/circuitplayground_express_crickit/mpconfigboard.mk]: I will re-test and let you know.

[the remark this answers is not in the document]
Comment (rhooper) [reply ports/atmel-samd/boards/circuitplayground_express_crickit/mpconfigboard.mk]: Seems to be 3760 bytes on metro_m4 after the changes (about to be pushed).  I can't currently build circuitplayground_express and circuitplayground_express_crickit locally, so I'm going to see if Travis can without this disabled for circuitplayground_express_crickit.

[the remark this answers is not in the document]
Comment (rhooper) [reply ports/atmel-samd/boards/circuitplayground_express_crickit/mpconfigboard.mk]: with _pixelbuf, circuitplayground_express_crickit, without changing the frozen libs:

```88 bytes free in flash out of 253440 bytes ( 247.5 kb ).
25868 bytes free in ram for stack out of 32768 bytes ( 32.0 kb ).```

and without

```4232 bytes free in flash out of 253440 bytes ( 247.5 kb ).
25876 bytes free in ram for stack out of 32768 bytes ( 32.0 kb ).
```
